Review precompiles benchmarks to display mgas/s

ahamlat opened the pull request:
## PR description
Review precompiles benchmarking to have an output like this for all the precompiles and Keccak.
```
...
BNMUL for 6,000 gas. 	Execution time per call 43,646 ns. 	@ 137.47 MGps
BNPairings 2 for 79,000 gas. 	Execution time per call 684,177 ns. 	@ 115.47 MGps
BNPairings 4 for 113,000 gas. 	Execution time per call 927,219 ns. 	@ 121.87 MGps
...
```
This is still a draft, so I haven't completed all the precompiles yet. I'll also do some refactoring to eliminate code duplication and improve the readability of the output through better indentation, to have something like 

```
...
BNMUL        for   6,000 gas. 	Execution time per call  43,646 ns. 	@ 137.47 MGps
BNPairings 2 for  79,000 gas. 	Execution time per call 684,177 ns. 	@ 115.47 MGps
BNPairings 4 for 113,000 gas. 	Execution time per call 927,219 ns. 	@ 121.87 MGps
...
```


## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->


### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] spotless: `./gradlew spotlessApply`
- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

garyschulte reviewed (commented):
This output format is more intuitive and useful imo.  I think if we need to try to derive gas costs/targets, we can still do so.

ahamlat commented on evm/src/test/java/org/hyperledger/besu/evm/precompile/Benchmarks.java:
```diff
@@ -41,10 +41,14 @@
 @SuppressWarnings("UnusedMethod")
 public class Benchmarks {
 
+  private static final Logger LOG = LoggerFactory.getLogger(Benchmarks.class);
```
Yes, it is not yet implemented

ahamlat commented on evm/src/test/java/org/hyperledger/besu/evm/precompile/Benchmarks.java:
```diff
@@ -609,16 +639,35 @@ private static double runBenchmark(final Bytes arg, final PrecompiledContract co
-    benchSha256();
-    benchKeccak256();
-    benchRipeMD();
+    // benchSecp256k1Recover();
+    // benchSha256();
+    // benchKeccak256();
+    // benchRipeMD();
```
Yes, I wanted to have a first feedback to know if it is worth pursuing the changes.

ahamlat commented:
@shemnon What do you think of using actual execution time and the gas requirement to get the mgas/s, instead of deriving a theoretical gas cost from execution time, with the assumption that 35 mgas should take 1 second ?
I guess this was a way to detect if the pricing is fair for precompiles, at least on besu side. I would like to change it to display a report of all the precompiles performances. What do you think ?

shemnon commented:
If we are _not_ trying to determine what a precompile _should_ cost then the MGps is a fine metric.  But if the gas schedule is up for debate I found the standard burn rate a reasonable proxy.

35 Mgps was an old 2020 era standard and I think we are starting to get pressure for 100 MGps.  But the keccak opcode is going to hold us back.

ahamlat commented:
I think we can keep both, then we can use the output for both use cases. I will update the PR. I agree on 100 mgas/s, I will update to that target.

ahamlat commented:
I updated the description, and added the output file below.
[PR8495-output.zip](https://github.com/user-attachments/files/19566032/PR8495-output.zip)

siladu commented on evm/src/test/java/org/hyperledger/besu/evm/precompile/Benchmarks.java:
```diff
@@ -32,25 +25,19 @@
-
-  static final long GAS_PER_SECOND_STANDARD = 35_000_000L;
-
-  static final int HASH_WARMUP = 1_000_000;
```
Is the 10_000 MATH_WARMUP sufficient for all cases?

ahamlat replied:
We can set it to to 15_000 to be sure to hit tier4 of c2 compilation. This should be sufficient for keccak as it is all Java. I don't know why the other threshold exists for Rimped and sha256. It was there since the creation of this file.
```
java -XX:+PrintFlagsFinal -version | grep CompileThreshold
     intx CompileThreshold                         = 10000                                  {pd product} {default}
   double CompileThresholdScaling                  = 1.000000                                  {product} {default}
    uintx IncreaseFirstTierCompileThresholdAt      = 50                                        {product} {default}
     intx Tier2CompileThreshold                    = 0                                         {product} {default}
     intx Tier3CompileThreshold                    = 2000                                      {product} {default}
     intx Tier4CompileThreshold                    = 15000                                     {product} {default}

```

siladu reviewed (approved):
LGTM
